File: examples/lua/ws_client.lua
-- hv.ws coroutine-synchronous WebSocket client demo.
-- Usage: hvlua examples/lua/ws_client.lua [url]
local url = arg[1] or "ws://127.0.0.1:8888/"

hv.setTimeout(1, function()
    local ws, err = hv.ws.connect(url)
    if err then
        hv.loge("connect failed:", err)
        hv.stop()
        return
    end
    hv.log("connected to", url)

    ws:send("hello from lua")

    -- recv() suspends the coroutine until a message arrives (or the peer closes,
    -- in which case it returns nil, "closed"). Loop until closed.
    for i = 1, 3 do
        local msg, rerr = ws:recv()
        if rerr then hv.log("recv:", rerr); break end
        hv.log("recv ->", msg)
        ws:send("echo " .. i)
    end

    ws:close()
    hv.stop()
end)

File: examples/js/ws_client.js
// hv/ws Promise WebSocket client demo.
// Usage: hvjs examples/js/ws_client.js [url]

const hv = require("hv");
const wsmod = require("hv/ws");

const url = arg[1] || "ws://127.0.0.1:8888/";
const ws = await wsmod.connect(url);

hv.log("connected to", url);
ws.send("hello from js");

for (let i = 1; i <= 3; ++i) {
    const msg = await ws.recv();
    hv.log("recv ->", msg);
    ws.send("echo " + i);
}

ws.close();
